### PR TITLE
Implement syntax and structure for term name/select/apply/applytype/this

### PR DESCRIPTION
**Pull request by olafurpg**
I used this as a warming up exercise to become familiar with the codebase and how to unapply/apply abstract syntax. It doesn't feel too different from the converter-based approach in scalameta/paradise.

The pretty-printers are implemented in a very raw way, everything inlined with a lot of duplicated code. I wanted to get the basic tests passing first before trying to abstract over the boilerplate. However, I don't expect TreeSyntax or TreeStructure will become much more complicated to merit sophisticated pretty-printing combinators.

The newly added structure/syntax tests are a good foundation to test the portability of very simple macros. Next up is to get some def macros working in Dotty 😎

### CODE REVIEW

File: engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala [hunk 1]
```diff
@@ -344,7 +370,13 @@ trait Abstracts extends scala.macros.trees.Abstracts with Positions { self: Univ
...
       }
       def unapply(gtree: Any): Option[String] = gtree match {
-        case g.Ident(name: g.TypeName) => unapply(name.decoded)
+        case g.Ident(name: g.TypeName) => Some(name.decoded)
+        case tt @ g.TypeTree()
+            if tt.original == null &&
+              tt.symbol != null &&
+              !tt.symbol.info.isHigherKinded =>
+          // TODO(olafur) we may want to emit Type.Select instead of Type.Name here.
+          Some(tt.symbol.decodedName)
```
Comment (liufengyun) [inline engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala:385]: As we discussed before, extractors for typed trees are better to be delayed. In gestalt, we've seen no use of them at all. Working on types instead of type trees or strings is more robust and leads to much less error-prone.

File: engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala [hunk 2]
```diff
@@ -374,7 +406,16 @@ trait Abstracts extends scala.macros.trees.Abstracts with Positions { self: Univ
...
+      def unapply(gtree: Any): Option[(Type, List[Type])] = gtree match {
+        // TODO(olafur) hack, shady stuff happening here.
+        case tt @ g.TypeTree()
+            if tt.original == null &&
+              tt.symbol != null &&
+              tt.symbol.info != null &&
+              tt.symbol.info.takesTypeArgs =>
+          Some(g.TypeTree(tt.symbol.info) -> tt.symbol.info.typeArgs.map(g.TypeTree))
+        case _ => None
+      }
```
Comment (liufengyun) [inline engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala:425]: The same as above, we delay extractors for type trees. Instead, we can introduce extractors for commonly used types. We can do in another PR.

File: tests/macros/src/main/scala/scala/macros/tests/scaladays/TestMacros.scala [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package scala.macros.tests.scaladays
+
+import scala.macros._
+
+import scala.language.experimental.macros
```
Comment (liufengyun) [inline tests/macros/src/main/scala/scala/macros/tests/scaladays/TestMacros.scala]: I guess this line is only temporary, will eventually be removed?

File: tests/macros/src/test/scala/scala/macros/tests/scaladays/StructureSuite.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  @Test def apply1(): Unit =
+    assertEquals(
+      """Term.Apply(Term.Select(Lit.String("a"), Term.Name("charAt")), List(Lit.Int(0)))""",
+      structure("a".charAt(0))
+    )
+  @Test def apply2(): Unit =
+    assertEquals(
+      "Term.Apply(Term.Select(Lit.String(\"a\"), Term.Name(\"substring\")), List(Lit.Int(0), Lit.Int(1)))",
+      structure("a".substring(0, 1))
+    )
```
Comment (liufengyun) [inline tests/macros/src/test/scala/scala/macros/tests/scaladays/StructureSuite.scala:42]: This reminds me of testing the `platform-independent quasiquotes` in gestalt: https://github.com/liufengyun/gestalt/blob/master/macros/src/main/scala/gestalt/macros/Quasiquotes.scala#L23 The tests are a little annoying, as they have to be wrapped in macros and defined in another project.

We'll be able to use the structure prettyprint to test `platform-independent quasiquotes`, which is nice.

File: tests/macros/src/test/scala/scala/macros/tests/scaladays/SyntaxSuite.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  @Test def applyCurry(): Unit =
+    assertEquals("SyntaxSuite.this.add(1)(2)", syntax(add(1)(2)))
+
+  def tpe[T](e: T) = e
+  @Test def typeApply(): Unit =
+    assertEquals("SyntaxSuite.this.tpe[Int](2)", syntax(tpe(2)))
+  val lst = List(1)
+  @Test def typeApplyHK(): Unit = {
+    assertEquals("SyntaxSuite.this.tpe[List[Int]](SyntaxSuite.this.lst)", syntax(tpe(lst)))
+  }
```
Comment (liufengyun) [inline tests/macros/src/test/scala/scala/macros/tests/scaladays/SyntaxSuite.scala:56]: Prettyprint the syntax is very helpful in debugging macros.

But if implemented in the macros abstractly, they will be limited by extractors provided, which may become an obstacle for debugging. I think a design comprise could be use the host prettyprint for syntax.

File: tests/macros/src/test/scala/scala/macros/tests/scaladays/SerializeSuite.scala [hunk 6]
```diff
@@ -5,14 +5,12 @@ import org.junit._
...
-import scala.macros.config._
 
 @RunWith(classOf[JUnit4])
 class SerializeSuite {
   @Test
   def simple: Unit = {
     case class C(x: Int, y: String)
-    def serialize[T: Serialize](x: T): String = implicitly[Serialize[T]].apply(x)
-    assertEquals("""{ "x": 40, "y": "2" }""", serialize(C(40, "2")))
+    assertEquals("""{ "x": 40, "y": "2" }""", Serialize.materialize[C](C(40, "2")))
```
Comment (liufengyun) [inline tests/macros/src/test/scala/scala/macros/tests/scaladays/SerializeSuite.scala:14]: Nice, I like semantic tests.

File: engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala [hunk 7]
```diff
@@ -95,47 +95,63 @@ trait Abstracts extends scala.macros.trees.Abstracts with Positions { self: Univ
...
     }
 
     object TermThis extends TermThisCompanion {
-      def apply(qual: Name): Term.Ref = ???
-      def unapply(gtree: Any): Option[Name] = ???
+      def apply(qual: Name): Term = g.This(qual.name.toTypeName)
+      def unapply(gtree: Any): Option[Name] = gtree match {
+        case g.This(g.TypeName(name)) => Some(c.NameIndeterminate(name))
+        case _ => None
+      }
```
Comment (liufengyun) [inline engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala:154]: Do we delay extractors for `This` and `Super`? We can address this in another PR with the split of typed and untyped trees, I guess at least all extractors have to be typed.

File: core/src/main/scala/scala/macros/internal/trees/TreeStructure.scala [hunk 8]
```diff
@@ -6,8 +6,80 @@ import scala.macros.Universe
...
+        p.raw(x.syntax)
+        p.raw(")")
+      case Lit.Symbol(_) =>
+        p.raw("Lit.Symbol(")
+        p.raw(x.syntax)
+        p.raw(")")
+      case _ =>
+        println(s"NotImplementedYet: $x (${x.getClass})")
+        ???
+    }
```
Comment (liufengyun) [inline core/src/main/scala/scala/macros/internal/trees/TreeStructure.scala:83]: My experience is that during debugging, I only check structure when I'm trying to fix bugs in quasiquote implementation (macro system dev). Otherwise, if I'm developing macros (macro dev), most of the time I use syntax for debugging.

The design philosophy of restricting number of extractors is in tension with implementing pretty-printing abstractly. A potential solution is to delegate the implementation to backends. And in the beginning, backends can just print the tree. When there's compelling need to converge the output, it's possible to do so as well.

File: engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala [hunk 9]
```diff
@@ -181,12 +200,19 @@ trait Abstracts extends scala.macros.trees.Abstracts with Positions { self: Univ
 
     object TermApply extends TermApplyCompanion {
       def apply(fun: Term, args: List[Term]): Term = g.Apply(fun, args)
-      def unapply(gtree: Any): Option[(Term, List[Term])] = ???
+      def unapply(gtree: Any): Option[(Term, List[Term])] = gtree match {
+        case gtree @ g.Apply(fun, args) if LitSymbol.symApply != gtree.symbol =>
```
Comment (liufengyun) [inline engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala]: I don't understand the guard `if LitSymbol.symApply != gtree.symbol `. I guess it's something scalac-specific? Just curious.

**Review by liufengyun — APPROVED** [review]
Good job @olafurpg , LGTM.

I added some related thoughts, but most of them are for latter PRs and open-ended discussion.

File: engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala (hunk 7 above)
Comment (olafurpg) [reply engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala:154]: I implemented this extractor as part of my exercise to support the `syntax(x) == "SyntaxSuite.this.x"` test. I felt like I needed better understanding of the syntax-based approach before embarking on the separation of typed and untyped trees.

File: engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala (hunk 9 above)
Comment (olafurpg) [reply engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala]: This was to support the test `syntax('S) == "'S"`, scalac desugars it into `scala.Symbol.apply("S")`. I've added a note saying we may want to skip this resugaring later down the road, this was just an experiment to support symmetric structure for literals.

File: engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala (hunk 1 above)
Comment (olafurpg) [reply engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala:385]: I agree. This was simply an experiment to see how hard it was to support for syntax/structure. I've added a note to either remove this or make it private.

File: engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala (hunk 2 above)
Comment (olafurpg) [reply engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/trees/Abstracts.scala:425]: Agreed, added note.

File: tests/macros/src/main/scala/scala/macros/tests/scaladays/TestMacros.scala (hunk 3 above)
Comment (olafurpg) [reply tests/macros/src/main/scala/scala/macros/tests/scaladays/TestMacros.scala]: I've added `-language:experimental.macros` to scalacOptions to avoid the need for this line.  The scalac-plugin SyntaxAnalyzer inserts `import scala.language.exprimental.macros` but it doesn't seem to be picked up by   scalac for all macros, only the Serialize macro.

File: tests/macros/src/test/scala/scala/macros/tests/scaladays/SerializeSuite.scala (hunk 6 above)
Comment (olafurpg) [reply tests/macros/src/test/scala/scala/macros/tests/scaladays/SerializeSuite.scala:14]: The `syntax(a.b) == "a.b"` tests are strictly speaking just as much semantic tests 😜 

I agree however that we should focus on "real-world" macros instead of synthetic cases. This PR was really only for me to understand things better.

File: tests/macros/src/test/scala/scala/macros/tests/scaladays/StructureSuite.scala (hunk 4 above)
Comment (olafurpg) [reply tests/macros/src/test/scala/scala/macros/tests/scaladays/StructureSuite.scala:42]: Wow, nice test suite! I hadn't seen that one

File: tests/macros/src/test/scala/scala/macros/tests/scaladays/SyntaxSuite.scala (hunk 5 above)
Comment (olafurpg) [reply tests/macros/src/test/scala/scala/macros/tests/scaladays/SyntaxSuite.scala:56]: I agree, for this last test I in fact use Tree.toString to get `List[Int]`. The way I see the syntax/structure tests is that they document the extent of what macro authors are able to pattern match against.

**Review by olafurpg — COMMENTED** [review]
Thank you for the review @liufengyun. I've addressed your comments.
